prost-types/src/datetime.rs

```diff
@@ -0,0 +1,756 @@
+    pub(crate) hour: u8,
+    /// The minute of the hour, from 0 to 59, inclusive.
+    pub(crate) minute: u8,
+    /// The second of the minute, from 0 to 59, inclusive.
+    pub(crate) second: u8,
+    /// The nanoseconds, from 0 to 999_999_999, inclusive.
+    pub(crate) nanos: u32,
+}
+
+impl DateTime {
+    /// The minimum representable [`Timestamp`] as a `DateTime`.
+    pub(crate) const MIN: DateTime = DateTime {
+        year: -292_277_022_657,
+        month: 1,
+        day: 27,
+        hour: 8,
+        minute: 29,
+        second: 52,
+        nanos: 0,
+    };
+
+    /// The maximum representable [`Timestamp`] as a `DateTime`.
+    pub(crate) const MAX: DateTime = DateTime {
+        year: 292_277_026_596,
+        month: 12,
+        day: 4,
+        hour: 15,
+        minute: 30,
+        second: 7,
+        nanos: 999_999_999,
+    };
+
+    /// Returns `true` if the `DateTime` is a valid calendar date.
+    pub(crate) fn is_valid(&self) -> bool {
+        self >= &DateTime::MIN
+            && self <= &DateTime::MAX
+            && self.month > 0
+            && self.month <= 12
+            && self.day > 0
+            && self.day <= days_in_month(self.year, self.month)
+            && self.hour < 24
+            && self.minute < 60
+            && self.second < 60
+            && self.nanos < 1_000_000_000
+    }
+}
+
+impl fmt::Display for DateTime {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // Pad years to at least 4 digits.
+        if self.year > 9999 {
+            write!(f, "+{}", self.year)?;
+        } else if self.year < 0 {
+            write!(f, "{:05}", self.year)?;
+        } else {
+            write!(f, "{:04}", self.year)?;
+        };
+
+        write!(
+            f,
+            "-{:02}-{:02}T{:02}:{:02}:{:02}",
+            self.month, self.day, self.hour, self.minute, self.second,
+        )?;
+
+        // Format subseconds to either nothing, millis, micros, or nanos.
+        let nanos = self.nanos;
+        if nanos == 0 {
+            write!(f, "Z")
+        } else if nanos % 1_000_000 == 0 {
+            write!(f, ".{:03}Z", nanos / 1_000_000)
+        } else if nanos % 1_000 == 0 {
+            write!(f, ".{:06}Z", nanos / 1_000)
+        } else {
+            write!(f, ".{:09}Z", nanos)
+        }
+    }
+}
+
+impl From<crate::Timestamp> for DateTime {
+    /// musl's [`__secs_to_tm`][1] converted to Rust via [c2rust][2] and then cleaned up by hand.
+    ///
+    /// All existing `strftime`-like APIs in Rust are unable to handle the full range of timestamps
+    /// representable by `Timestamp`, including `strftime` itself, since tm.tm_year is an int.
+    ///
+    /// [1]: http://git.musl-libc.org/cgit/musl/tree/src/time/__secs_to_tm.c
+    /// [2]: https://c2rust.com/
+    fn from(timestamp: crate::Timestamp) -> DateTime {
+        let t = timestamp.seconds;
+        let nanos = timestamp.nanos;
+
+        // 2000-03-01 (mod 400 year, immediately after feb29
```

Comment:
Missing `)`?

Reply:
This typo exists in the original C version: https://git.musl-libc.org/cgit/musl/tree/src/time/__secs_to_tm.c#n4 .  I've never fixed it because I'm not exactly sure where the closing paren should go.

Reply:
lol